feat(compression): update tooling to use DECODE operators

rkuester opened the pull request:
This is a draft PR for running CI tests on the full change. The commits
along this branch will be individually submitted for review.

See the linked issue for a description of the change.

BUG=implements #3256

rkuester committed:
test(compression): add integration tests with TFLM interpreter

Add tests that compress models with LUT compression, run them through
the TFLM Python interpreter, and verify outputs match uncompressed
originals. Also verify DECODE operators are inserted and that compressed
models are smaller than originals.

Tests only run when compression is enabled (--//:with_compression).
Placeholder tests for Huffman and Pruning are skipped until implemented.

rkuester committed:
refactor(compression): extract _parse_compression_method helper

Factor out compression method parsing into a dedicated function that
dispatches on the YAML key. This enables parse_yaml to iterate over
multiple compression methods per tensor and makes adding new
compression types straightforward.

rkuester committed:
feat(compression): implement model_editor for TFLite model manipulation

Implement unified module for creating, reading, and modifying TFLite
models with a clean API. The module eliminates manual index tracking
and buffer management through automatic bookkeeping, supporting both
declarative and imperative construction styles.

Core design uses first-class Buffer objects that can be shared between
tensors, with automatic deduplication during build. Tensors reference
Buffers directly, matching the TFLite schema structure. The compiler
automatically extracts inline tensor declarations, builds operator code
tables, and handles index assignment according to TFLite conventions.

Supports quantization parameters (per-tensor and per-channel), metadata
key-value pairs, and read-modify-write workflows. The read() function
preserves the object graph structure, enabling models to be read,
modified, and rebuilt.

Add comprehensive test coverage for core functionality, advanced
features, quantization, and modification workflows.

rkuester committed:
docs(compression): add docstring to LookUpTableCompression

rkuester committed:
feat(compression): add LUT compression plugin

Extract LUT compression logic from compress.py into a dedicated plugin
module. The LutCompressor class implements the Compressor protocol,
producing packed indices and ancillary data in the format expected by
the C++ DECODE kernel.

rkuester committed:
feat(compression): add DECODE operator insertion logic

Implement graph modification to insert DECODE operators before
consumers of compressed tensors. Each compressed tensor gets a DECODE
operator with two inputs (encoded tensor and ancillary data tensor)
and one output (decompressed tensor). Consumer operators are rewired
to use the DECODE output.

rkuester committed:
feat(compression): add Huffman and Pruning plugin stubs

Add placeholder implementations that raise CompressionError when
invoked. These validate the plugin architecture and will be replaced
with working implementations later.

rkuester committed:
feat(model_editor): add subgraph inputs and outputs fields

The TFLM interpreter requires subgraph inputs/outputs to be set in the
flatbuffer to know which tensors are model inputs and outputs. Without
these, models built with model_editor cannot be executed.

Add inputs and outputs fields to Subgraph dataclass, populate them in
_compile_subgraph when building, and read them back in read().

rkuester committed:
refactor(compression): migrate compress.py from model_facade to model_editor

Replace model_facade with model_editor in compress.py and tests.
model_editor provides a cleaner API with better buffer and metadata
handling. Buffers appended during compression are automatically indexed,
and quantization parameters are accessed through a wrapper object.

Update BUILD dependencies accordingly.

rkuester committed:
refactor(compression): use plugin architecture in compress.py

Replace monolithic compression logic with a dispatch table that routes
compression requests to plugin modules based on the spec's compression
method type. After compressing tensors, insert DECODE operators into
the model graph.

The old metadata flatbuffer approach is removed in favor of the DECODE
operator format.

rkuester committed:
feat(compression): add Compressor protocol and CompressionResult

Define the plugin interface for compression methods. Each compressor
implements the Compressor protocol with a compress() method that
returns encoded data and ancillary data. CompressionError provides
a common exception type for compression failures.

rkuester committed:
chore(compression): remove test_models.py and test_models_test.py

Remove test_models module and its tests as they are superseded by
model_editor.

rkuester committed:
refactor(compression): replace test_models with model_editor in compress_test

Replace dictionary-based test_models.build() with model_editor's
declarative API. Add _build_test_model() function that uses model_editor
to create the same test model more cleanly.

rkuester committed:
feat(compression): add HuffmanCompression and PruningCompression types

Add dataclass placeholders for future compression methods. These will
be used by the plugin architecture to dispatch to compression-specific
implementations.

rkuester committed:
feat(compression): add DecodeType class for type-safe decode operations

Add DecodeType class to replace raw integer decode_type field with
named constants and factory methods. Provides predefined constants for
built-in types (LUT, HUFFMAN, PRUNING) and a factory method for custom
types (128-255).

Custom types are automatically named with CUSTOM_{code} prefix for
clarity in debugging. The class supports serialization via __int__()
and comparison with both DecodeType objects and integers.

Update DecodeCommonMetadata to use DecodeType and update tests to use
named constants.

Add decode module BUILD targets.

rkuester committed:
chore(compression): remove model_facade.py and model_facade_test.py

Remove model_facade module and its tests as they are superseded by
model_editor.

rkuester committed:
feat(python): add alt decompression memory parameter to interpreter

Add alt_decompression_memory_size parameter to the Python interpreter
API. When non-zero, allocates a separate memory region for DECODE
operator outputs and calls SetDecompressionMemory before AllocateTensors.

SetDecompressionMemory stores a pointer to its initializer_list
argument, requiring the list to outlive the interpreter. Per C++
standard, an initializer_list's backing array lifetime is only extended
to match the list's when initialized in a declaration, not when
assigned. This makes the API difficult to use correctly.

rkuester committed:
test(compression): add alt decompression memory integration test

Add test for shared compressed tensors with alternate decompression
memory. The test is marked expectedFailure to document the current
mismatch between interpreter and DECODE insertion: the interpreter's
alt decompression memory resets allocations for each DECODE, but the
insertion code shares one DECODE output among all consumers.

The workaround is to insert a separate DECODE before each consumer.
The expectedFailure decorator should be removed once this is
implemented.

rkuester committed:
fix(compression): insert DECODE per consumer for alt decompression memory

Insert a separate DECODE immediately before each consumer of a
compressed tensor, rather than sharing one DECODE output among all
consumers.

The interpreter's alternate decompression memory resets its allocation
offset for each DECODE's Prepare, causing all DECODE outputs to be
allocated at the same address. If two consumers share one DECODE and
another DECODE runs between them, the intervening DECODE overwrites the
shared output, corrupting data for the second consumer.

Update test expectations to reflect the new DECODE-per-consumer
behavior and change the integration test from expected-failure to
expected-pass.

ddavis-2015 commented on tensorflow/lite/micro/compression/decode_insert.py:
```diff
@@ -0,0 +1,227 @@
+          info.ancillary_data,
+          info.tensor,
+      )
+      subgraph.tensors.append(ancillary_tensor)
```
This appears to create too many ancillary tensor copies.  Each encoded tensor can reuse the associated ancillary tensor.  Should _create_ancillary_tensor have a cache keyed on the original tensor?

ddavis-2015 commented on tensorflow/lite/micro/compression/decode_insert.py:
```diff
@@ -0,0 +1,227 @@
+      subgraph.operators.insert(insert_pos, decode_op)
+
+      # Rewire only this consumer to use the decoded output
+      _rewire_consumers([consumer], info.tensor, output_tensor)
```
The code to comply with "Input tensor pairs" and "Tensor rewriting" in the design doc seem to be missing.  Should there be another loop after the loop with _rewire_consumers that iterates compression_results and rewrites the original tensor (removing quantization, setting type to UINT8, changing shape to single rank and equal to the number of bytes in the encoded tensor)?

rkuester committed:
test(model_editor): add expected-failure tests for read() edge cases

Add tests demonstrating bugs in model_editor.read() when parsing models
with None values for tensor shape, operator inputs, or operator outputs.
These edge cases can occur in real models from the TFLite converter but
cause TypeError crashes in the current implementation.

Tests construct models using the low-level TFLite schema to reproduce
these conditions. Marked as expectedFailure until the fix is applied.

rkuester committed:
fix(model_editor): handle None shape and inputs/outputs in read()

The TFLite flatbuffer schema allows None values for tensor shape
(representing scalars) and operator inputs/outputs (for certain ops).
Handle these cases in read() to avoid TypeError when iterating.

Remove expectedFailure decorators from edge case tests now that the
fix is applied.

ddavis-2015 commented on tensorflow/lite/micro/compression/model_editor.py:
```diff
@@ -0,0 +1,572 @@
+          # Handle axis: only set if per-channel (more than one scale)
+          axis = None
+          if len(scales) > 1 and fb_quant.quantizedDimension is not None:
+            axis = fb_quant.quantizedDimension
```
The axis should always be copied if it exists.  It is perfectly valid for len(scales) to be 1, the quantized_dimension to be 3, and shape[3] is equal to 1.

rkuester replied:
Addressed. The [quantization parsing logic](https://github.com/tensorflow/tflite-micro/blob/1e5651fa/tensorflow/lite/micro/compression/model_editor.py#L562-L568) now preserves the axis whenever it's non-zero, regardless of scale count. Only the case of axis==0 with a single scale is treated as per-tensor quantization (axis=None). This preserves per-channel quantization with a single channel at a non-zero axis.

ddavis-2015 commented on python/tflite_micro/interpreter_wrapper.cc:
It occurs to me there is no reason to include alternate decompression memory APIs for the Python MicroInterpreter.  Not sure why I didn't bring this up before...

ddavis-2015 commented on python/tflite_micro/runtime.py:
It occurs to me there is no reason to include alternate decompression memory APIs for the Python MicroInterpreter.  Not sure why I didn't bring this up before...

ddavis-2015 commented on tensorflow/lite/micro/compression/compress.py:
```diff
@@ -293,46 +121,50 @@ def _apply_flatbuffer_alignment(model_bytes: bytearray) -> bytearray:
+      result = plugin.compress(tensor, method)
+
+      # Replace tensor data with encoded data
+      tensor.buffer.data = result.encoded_data
```
Could also occur later, after the DECODE insertion, when the rest of the tensor needs to be updated (removing quantization, changing shape and type, etc.)

ddavis-2015 commented on tensorflow/lite/micro/compression/pruning.py:
```diff
@@ -0,0 +1,59 @@
+from tflite_micro.tensorflow.lite.micro.compression import spec
+
+
+class PruningCompressor:
```
This does not inherit from Compressor?  Am I missing something here?

ddavis-2015 commented on tensorflow/lite/micro/compression/compress.py:
```diff
@@ -293,46 +121,50 @@ def _apply_flatbuffer_alignment(model_bytes: bytearray) -> bytearray:
+
+      method = tensor_spec.compression[0]
+      plugin = _get_compressor(method)
+      result = plugin.compress(tensor, method)
```
Warn user if this resulted in expansion (sizeof encoded_data + sizeof ancillary_data > original tensor length)?

rkuester replied:
Addressed. The code now [warns if compression results in expansion](https://github.com/tensorflow/tflite-micro/pull/3400/files#diff-1ca95116d18953c4b50538524c131a4443fe18af60a5d351288f567710dea6d1R155-R161).

ddavis-2015 commented on tensorflow/lite/micro/compression/lut.py:
```diff
@@ -0,0 +1,317 @@
+  return bytes(buffer)
+
+
+class LutCompressor:
```
This does not inherit from Compressor? Am I missing something here?

ddavis-2015 commented on tensorflow/lite/micro/compression/compress_test.py:
```diff
@@ -358,58 +152,199 @@ def test_multiple_tables_with_padding(self):
+        spec.Tensor(
+            subgraph=0,
+            tensor=0,
+            compression=[spec.LookUpTableCompression(index_bitwidth=4)])
```
index_bitwidth=2?

rkuester replied:
The tests now include both scenarios. Some tests use minimal bitwidth (like [`index_bitwidth=2` at lines 142-155](https://github.com/tensorflow/tflite-micro/pull/3400/files#diff-defc9e243c4ac63bb60953f79e2a1f6e6bce73d158f9e3f085da8f9dc0568ae3R142-R155)), while others intentionally use larger-than-necessary bitwidth to verify the warning behavior.

ddavis-2015 commented on tensorflow/lite/micro/compression/decode_insert.py:
```diff
@@ -0,0 +1,227 @@
+          custom_code=DECODE_CUSTOM_OP_NAME,
+          inputs=[info.tensor, ancillary_tensor],
+          outputs=[output_tensor],
+      )
```
This being located here does not allow for a single DECODE operator to have multiple encoded inputs and ancillary tensors.  The example would be CONCATENATION which takes multiple inputs, where some might be encoded tensors.

ddavis-2015 commented on tensorflow/lite/micro/compression/decode_insert_test.py:
```diff
@@ -0,0 +1,393 @@
+      shape=(1, 4),
+      dtype=tflite.TensorType.INT8,
+      name="output2",
+  )
```
Comments to label what number each tensor is.

rkuester replied:
See earlier reply about using `tensor_by_name()` to avoid magic tensor indices.

ddavis-2015 commented on tensorflow/lite/micro/compression/decode_insert_test.py:
```diff
@@ -0,0 +1,393 @@
+      shape=(1, 4),
+      dtype=tflite.TensorType.INT8,
+      name="output",
+  )
```
Comments to label what number each tensor is (or make it part of the tensor name, like T#5 perhaps?)

ddavis-2015 commented on tensorflow/lite/micro/compression/huffman.py:
```diff
@@ -0,0 +1,60 @@
+from tflite_micro.tensorflow.lite.micro.compression import spec
+
+
+class HuffmanCompressor:
```
Not a subclass of Compressor?  Am I missing something here?

ddavis-2015 commented on tensorflow/lite/micro/compression/model_editor.py:
```diff
@@ -0,0 +1,572 @@
+  dtype: tflite.TensorType
+  buffer: Optional[Buffer] = None
+  quantization: Optional[Quantization] = None
+  name: Optional[str] = None
```
is_variable must be maintained

rkuester replied:
We reworked model_editor to wrap the backing `*T` types (`TensorT`, `OperatorT`, `ModelT`, etc.) rather than using pure dataclasses. When compiling, we reuse the wrapped object and only update fields we explicitly manage. This preserves all other fields, including those added by future schema changes.

For Tensor, see [model_editor.py L769-770](https://github.com/tensorflow/tflite-micro/pull/3400/files#diff-42847d6ce72f11e805ba31973cf09ae235e36fd7c68ded63d7afc5343801edc9R769-R770). Verified by [test_tensor_is_variable_preserved](https://github.com/tensorflow/tflite-micro/pull/3400/files#diff-6aab2520fc52bb39d7976aed22d7094e16cf319d710c4c874a3c67ee975e33afR1096).

The same approach applies to similar comments below.

ddavis-2015 commented on tensorflow/lite/micro/compression/model_editor.py:
```diff
@@ -0,0 +1,572 @@
+  """Operator code specification."""
+  builtin_code: tflite.BuiltinOperator
+  custom_code: Optional[str] = None
+  version: int = 1
```
must maintain `deprecated_builtin_code`, and it must interoperate as per the schema.fbs comments with `builtin_code`.

rkuester commented on tensorflow/lite/micro/compression/model_editor.py:
```diff
@@ -0,0 +1,572 @@
+          scales = list(fb_quant.scale)
+          zeros = list(
+              fb_quant.zeroPoint
+          ) if fb_quant.zeroPoint is not None else [0] * len(scales)
```
Addressed. Zero_points are now [copied as-is without expansion](https://github.com/tensorflow/tflite-micro/pull/3400/files#diff-42847d6ce72f11e805ba31973cf09ae235e36fd7c68ded63d7afc5343801edc9R559-R561). A single zero_point with multiple scales is preserved.

rkuester commented on python/tflite_micro/_runtime.cc:
Kept for now. The alt memory API is used in integration tests to exercise that code path, which caught a bug earlier in development where sharing DECODE outputs broke in the presence of alt memory.

rkuester commented on tensorflow/lite/micro/compression/compression_integration_test.py:
```diff
@@ -0,0 +1,397 @@
+
+if __name__ == "__main__":
+  os.environ["TF_CPP_MIN_LOG_LEVEL"] = "2"
+  os.environ["TF_ENABLE_ONEDNN_OPTS"] = "0"
```
Addressed. The [environment variables are now documented](https://github.com/tensorflow/tflite-micro/pull/3400/files#diff-e72dddf8cd66c93e81f64bff38fe7c9f4a5e0ada35a01ffb7a3204c42009ab3dR502-R505).

rkuester commented on tensorflow/lite/micro/compression/compression_integration_test.py:
```diff
@@ -0,0 +1,397 @@
+      shape=(1, rows),
+      dtype=tflite.TensorType.INT8,
+      name="output",
+  )
```
Rather than adding magic number comments, we added [`tensor_by_name()`](https://github.com/tensorflow/tflite-micro/pull/3400/files#diff-42847d6ce72f11e805ba31973cf09ae235e36fd7c68ded63d7afc5343801edc9R432) so tests can look up tensors by name instead of index.

rkuester commented on tensorflow/lite/micro/compression/compressor.py:
```diff
@@ -0,0 +1,80 @@
+  @property
+  def decode_type(self) -> decode.DecodeType:
+    """The DecodeType constant for this compression method."""
+    ...
```
Kept as `...`. The ellipsis is the conventional way to indicate abstract methods in Protocol classes per PEP 544.

rkuester commented on tensorflow/lite/micro/compression/decode.py:
```diff
@@ -0,0 +1,240 @@
+  """Protocol for objects that can serialize ancillary data."""
+
+  def to_bytes(self) -> bytes:
+    ...
```
Kept as `...`. The ellipsis is the conventional way to indicate abstract methods in Protocol classes per PEP 544.

rkuester commented on tensorflow/lite/micro/compression/decode_insert.py:
```diff
@@ -0,0 +1,227 @@
+
+    if not consumers:
+      # Tensor not used as input anywhere---no DECODE needed
+      continue
```
Addressed. The new PR checks if the tensor is a subgraph output (raises `NotImplementedError` explaining compressed subgraph outputs aren't yet supported) or warns if it has no consumers and isn't an output. See [decode_insert.py L180-197](https://github.com/tensorflow/tflite-micro/pull/3400/files#diff-07340f9b1e2dd8d5fd5f026e5c8074b7e5168f1a415c0b6e7046001ad9325012R180-R197).

rkuester commented on tensorflow/lite/micro/compression/model_editor.py:
```diff
@@ -0,0 +1,572 @@
+  operators: List[Operator] = field(default_factory=list)
+  inputs: List[Tensor] = field(default_factory=list)
+  outputs: List[Tensor] = field(default_factory=list)
+  name: Optional[str] = None
```
See above about wrapping `*T` types. Verified by [test_operator_debug_metadata_index_preserved](https://github.com/tensorflow/tflite-micro/pull/3400/files#diff-6aab2520fc52bb39d7976aed22d7094e16cf319d710c4c874a3c67ee975e33afR1180).

rkuester commented:
Closing this draft. It was the original full-change preview for the
DECODE-operator tooling (#3256): replacing the metadata-based
compression scheme with explicit DECODE custom operators, plus the
plugin architecture for LUT/Huffman/Pruning compressors.

Reworked since then. The current full change lives in #3400, with
individual commits going up for review one at a time, starting with
#3439. #3400 supersedes this branch entirely. Closing it just to keep
the open-PR list focused on what's actually moving toward merge.
